tests/test_decoders.py

```diff
@@ -0,0 +1,201 @@
+
+        # run
+        self.test_decoder.upgrade_legacy()
+
+        # assert
+        assert self.test_decoder._models == [self.test_decoder.layers]
+        assert self.test_decoder.version == expected_version
+
+    @patch('steganogan.decoders.nn.Sequential', autospec=True)
+    def test_upgrade_legacy_with_version_1(self, sequential_mock):
+        """The object must be the same and not changed by the method"""
+        # setup
+        decoder = decoders.BasicDecoder(1, 1)
+        expected = copy.deepcopy(decoder)
+
+        # run
+        decoder.upgrade_legacy()
+
+        # assert
+        assert decoder.__dict__ == expected.__dict__
+
+    def test_forward_1_layer(self):
+        """If there is only one layer it must be called with image as the only argument."""
+        # setup
+        layer1 = Mock(return_value=torch.Tensor([[5, 6], [7, 8]]))
+        self.test_decoder._models = [layer1]
+
+        # run
+        image = torch.Tensor([[1, 2], [3, 4]])
+        result = self.test_decoder.forward(image)
+
+        call_1 = call(torch.Tensor([[1, 2], [3, 4]]))
+
+        # assert
+        assert (result == torch.Tensor([[5, 6], [7, 8]])).all()
+        assert_called_with_tensors(layer1, [call_1])
+
+    def test_forward_more_than_2_layers(self):
+        """If there are more than 2 layers, they must be called adding data to each result"""
+        # setup
+        layer1 = Mock(return_value=torch.Tensor([[5, 6], [7, 8]]))
+        layer2 = Mock(return_value=torch.Tensor([[9, 10], [11, 12]]))
+        layer3 = Mock(return_value=torch.Tensor([[13, 14], [15, 16]]))
+        self.test_decoder._models = [layer1, layer2, layer3]
+
+        # run
+        image = torch.Tensor([[1, 2], [3, 4]])
+        result = self.test_decoder.forward(image)
+
+        # asserts
+        call_layer_1 = call(torch.Tensor([[1, 2], [3, 4]]))
+        call_layer_2 = call(torch.Tensor([[5, 6], [7, 8]]))
+        call_layer_3 = call(torch.Tensor([[5, 6, 9, 10], [7, 8, 11, 12]]))
+
+        assert_called_with_tensors(layer1, [call_layer_1])
+        assert_called_with_tensors(layer2, [call_layer_2])
+        assert_called_with_tensors(layer3, [call_layer_3])
+
+        assert (result == torch.Tensor([[13, 14], [15, 16]])).all()
+
+
+class TestDenseDecoder(TestCase):
+
+    class TestDecoder(decoders.DenseDecoder):
+        def __init__(self):
+            pass
+
+    def test_upgrade_legacy_without_version(self):
+        """Upgrade legacy must create self._models from conv1, conv2, conv3, conv4"""
+        # setup
+        test_decoder = self.TestDecoder() # instance an empty decoder
+        test_decoder.conv1 = Mock(return_value=torch.Tensor([[5, 6], [7, 8]]))
+        test_decoder.conv2 = Mock(return_value=torch.Tensor([[9, 10], [11, 12]]))
+        test_decoder.conv3 = Mock(return_value=torch.Tensor([[13, 14], [15, 16]]))
+        test_decoder.conv4 = Mock(return_value=torch.Tensor([[17, 18], [19, 20]]))
+
+        expected_models = [
```

Comment:
Do not define variables in advance.
Define the "expected" variables in the "asserts" block.